### PR TITLE
Implement core ops for references + fix core module documentation

### PR DESCRIPTION
**Pull request by sellibitze**
As discussed with @aturon I added implementations of various op traits for references to built-in types which was already suggested by the ops reform RFC.

The 2nd commit updates the module documentation of core::ops to fully reflect the recent change from pass-by-reference to pass-by-value and expands on the implications for generic code.

### CODE REVIEW

File: src/libcore/ops.rs [hunk 1]
```diff
@@ -96,6 +104,80 @@ pub trait Drop {
...
+            }
+        }
+    }
+}
+
+// implements binary operators "&T op U", "T op &U", "&T op &U"
+// based on "T + U" where T and U are expected `Copy`able
+macro_rules! forward_ref_binop {
+    (impl $imp:ident, $method:ident for $t:ty, $u:ty) => {
+        forward_ref_val_binop! { impl $imp, $method for $t, $u }
```
Comment (huonw) [inline src/libcore/ops.rs]: I wonder if these three macros could be inlined and just have 3 impl blocks in here.

Comment (sellibitze) [inline src/libcore/ops.rs]: @huonw Sure, that's doable. I've never used them anywhere except in the bottom macro. And I think it should save 12 non-empty lines of code. Do you want me to inline them? Personally, I don't have a preference.

Comment (huonw) [inline src/libcore/ops.rs]: Yeah, I think they should be inlined; thanks!

**Comment by sellibitze** [comment]
updated macros as suggested

**Comment by sellibitze** [comment]
I checked libcoretest/ops.rs and didn't see any tests for these kinds of implementations. Otherwise I would have added tests there already, too. But I think it might be a good idea to add tests to make sure that all the operators are doing what they are supposed to and that we don't have a "copy-&-paste" bug that, for example, an implementation of Mul in terms of + or something like that.

Any suggestions? @huonw @aturon

**Comment by aturon** [comment]
@sellibitze 

Thanks for this PR, and sorry for the slow review (it was a holiday weekend).

This looks great. My only request is to annotate with `#[unstable = "recently added, waiting for dust to settle"]`.

**Comment by sellibitze** [comment]
@aturon Done.

**Comment by aturon** [comment]
@sellibitze Somehow I missed the comment about testing before. I'm ok taking this PR as-is, but would also be delighted if you wanted to add some tests like the ones you suggested. Should I wait for those before sending to bors?

**Comment by sellibitze** [comment]
@aturon I don't yet have such tests implemented because I waited for feedback. So, it would take a while to get this done. I'm ok with adding them to this or a new PR.

**Comment by aturon** [comment]
@sellibitze OK, in that case I think let's move forward with the PR as it stands, and some tests can be brought in as a follow-up PR. (In general we like to land tests together with new code, but in a case like this it's totally fine.)

Thanks again for doing this!

**Pull request merged**